pkg/applicationserver/applicationserver.go

```diff
@@ -643,6 +673,24 @@ func (as *ApplicationServer) attemptDownlinkQueueOp(ctx context.Context, dev *tt
 	}
 }
 
+// initializeConfirmationRetriesConfig initializes the confirmation retries
+// configuration for the given downlink items.
+func (as *ApplicationServer) initializeConfirmationRetriesConfig(items []*ttnpb.ApplicationDownlink) {
+	maxRetries := as.config.Downlinks.ConfirmationConfig.DefaultRetryAttempts
+	if as.config.Downlinks.ConfirmationConfig.MaxRetryAttempts != 0 {
+		maxRetries = as.config.Downlinks.ConfirmationConfig.MaxRetryAttempts
+	}
+
+	for _, item := range items {
+		if item.Confirmed && item.ConfirmedRetry == nil {
+			item.ConfirmedRetry = &ttnpb.ApplicationDownlink_ConfirmedRetry{
+				Attempt:     1,
+				MaxAttempts: wrapperspb.UInt32(uint32(maxRetries)),
```

Comment:
I wouldn't initialize `max_attempts` and instead just build it dynamically in the nack handler.